### PR TITLE
Upgraded to ve/evs/1.0 and added in mail function

### PR DESCRIPTION
**Pull request by GwenChen-NOAA**
## Pull Request Testing ##

- [ ] Describe testing already performed for this Pull Request:</br>

- [ ] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

- [ ] Has the code been checked to ensure that no errors occur during the execution? **[Yes or No]**

- [ ] Do these updates/additions include sufficient testing updates? **[Yes or No]**

- [ ] Please complete this pull request review by **[Fill in date]**.</br>

## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

### CODE REVIEW

**Comment by ShelleyMelchior-NOAA** [comment]
Hi @GwenChen-NOAA , I will pick up this PR for you.

**Comment by ShelleyMelchior-NOAA** [comment]
@GwenChen-NOAA , just to be intentionally over explicit, can you list which ecf scripts I need to run? All 10 prep, stats, and plots ecf scripts?

**Comment by GwenChen-NOAA** [comment]
Yes, they all have been updated to use ve/evs/1.0.

**Comment by ShelleyMelchior-NOAA** [comment]
@GwenChen-NOAA , please remove the `ush/rtofs/__pycache__/` directory in your fork's `develop`.

**Comment by ShelleyMelchior-NOAA** [comment]
@GwenChen-NOAA , can you update all 10 ecf/ scripts, 4 jobs/ scripts, and 6 scripts/ scripts  to use _**HOMEevs**_ instead of _**EVS**_?  To be consistent across all EVS components, I'd like to see RTOFS using HOMEevs. I know it might seem like a nitpick, but this is something I think NCO will notice.

**Comment by GwenChen-NOAA** [comment]
@Shelley Melchior - NOAA Affiliate ***@***.***> , please use
this version to test for now.  I will be on leave next Monday to Wednesday
(May 22-24), and the RTOFS Cable Transport Analyses
<https://polar.ncep.noaa.gov/global/class-4/cable/> webpage stopped
running after two weeks of missing data.  Jason asks me to fix it, so I
will do that tomorrow.  I won't be able to do anything for EVSv1 until next
Thursday, and I don't want you to wait for me.

Thanks,
Gwen

**Comment by ShelleyMelchior-NOAA** [comment]
Thanks @GwenChen-NOAA . Yes, I can test without these changes. I will provide updates here and look for some updates to your fork's develop next Thursday.

**Comment by ShelleyMelchior-NOAA** [comment]
I ran the **_prep_** step and I think all looks OK. I compared my local COMOUT to the EVS parallel COMOUT.

log: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_rtofs_prep.o58326989`
output: `/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0`

**Comment by ShelleyMelchior-NOAA** [comment]
I have the cron entries all set up and I have the stats jobs set to point COMIN and COMINfcst to Perry's para as his prep archive is more complete. I think I still don't have things set up correctly, though. The ndbc stats job just ran and it didn't complete correctly.

output log file: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_rtofs_ndbc_stats.out.sv`
The ecf code: `/lfs/h2/emc/vpppg/noscrub/shelley.melchior/githubwkspc/forks/GC/EVS`

**Comment by ShelleyMelchior-NOAA** [comment]
@GwenChen-NOAA  -- please review outputs when you can.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/`

**Comment by GwenChen-NOAA** [comment]
@Shelley Melchior - NOAA Affiliate ***@***.***> , WCOSS2
switched machines today, and I don't see your outputs on Dogwood.  Please
let me know when they are available on Dogwood.

Thanks,
Gwen

**Comment by PerryShafran-NOAA** [comment]
Shelley is on leave until Thursday, so I would guess that she is not running them currently.  I haven't been following but perhaps I can pick it up (which would be tomorrow), or she can pick it up when she returns.

Perry

**Comment by GwenChen-NOAA** [comment]
Thanks, @Perry Shafran - NOAA Affiliate ***@***.***> !  Keep me
posted.

Gwen

**Comment by AliciaBentley-NOAA** [comment]
Hi Perry,

If you could pick this PR up while Shelley is out of the office, that would
be great! Thank you for your help!

Take care,
Alicia

___________________________________
*Alicia M. Bentley, Ph.D.*
Global Verification Project Lead
Model Evaluation Group (MEG) <https://www.emc.ncep.noaa.gov/users/meg/home/>
NOAA/NWS/NCEP/EMC (VPPPG Branch)
Work Phone: ‪(301) 683-3783‬

**Comment by PerryShafran-NOAA** [comment]
I first need to transfer a good deal of prep data from Cactus to Dogwood.  I do have a regular mirror from the dev machine to prod to prepare for switches, but apparently when Cactus got slow, so did the mirror, so the past 5 days did not get transferred.  I would need that prep data to do any testing anyway.  (This is true for other EVS components as well, such as subseasonal, and I need to check others.)  That's a lot of data to move around.  Stay tuned.

Perry

**Comment by PerryShafran-NOAA** [comment]
Per Alicia's request, I am picking this up.  I'm running the stats part now, and I've linked to the emc.vpppg rtofs prep archive (though I know I will need to test prep as well).

Perry

**Comment by ShelleyMelchior-NOAA** [comment]
@GwenChen-NOAA , @PerryShafran-NOAA , @AliciaBentley-NOAA , I am back from leave. This PR has lingered b/c of the unfortunate timing of developer leave, CM leave, and the WCOSS2 maintenance schedule. Perry will resume work on younger PRs and I will resume work on RTOFS. I will need a little bit of time to transfer my original set up from cactus to dogwood and I do not know how efficient dogwood is working today since many nodes are not available. Please be patient.

**Comment by AliciaBentley-NOAA** [comment]
Hi Shelley,

Thank you for picking up this PR again! It would be excellent if we could
get a handful of the nearly-complete/small PRs merged today as we begin the
EVSv1 Beta Test!

Take care,
Alicia
___________________________________
*Alicia M. Bentley, Ph.D.*
Global Verification Project Lead
Model Evaluation Group (MEG) <https://www.emc.ncep.noaa.gov/users/meg/home/>
NOAA/NWS/NCEP/EMC (VPPPG Branch)
Work Phone: ‪(301) 683-3783‬

**Comment by ShelleyMelchior-NOAA** [comment]
@GwenChen-NOAA , I got the environment set up on dogwood and have run the prep job interactively. You can view the output log file and prep file output directory (COMOUT) here:
log: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_rtofs_prep.out`
`COMOUT=/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0`

I also have it all set up in cron to run as the EVS parallel does. The stats processing is pointing to the EVS para prep files for COMIN. Prep will run again via cron at 0700. Stats don't run until 2200, so I can run those interactively and provide output for your to review.

**Comment by ShelleyMelchior-NOAA** [comment]
2 things for you to do before I can merge this PR after you approve the runs.

1. please update your feature branch to pull in all of the updates to the auth repo develop.
2. update ecf/, jobs/, and scripts/ to change $EVS usage to $HOMEevs.

**Comment by ShelleyMelchior-NOAA** [comment]
@GwenChen-NOAA  I ran the argo stats interactively. I do not think it ran correctly. Please review the log file to make sure I have all of the settings correct.
log: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_rtofs_argo_stats.out`
And if it helps, my HOMEevs=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/githubwkspc/forks/GC/EVS`

**Comment by GwenChen-NOAA** [comment]
@Shelley Melchior - NOAA Affiliate ***@***.***> , my argo
stats ran successfully yesterday.  In your jevs_rtofs_argo_stats.out file,
you have the following outputs:

DEBUG 1: Start gen_vx_mask by shelley.melchior(32081) at 2023-06-02
02:54:39Z  cmd: gen_vx_mask
/lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/prep/rtofs/rtofs.20230530/argo/
rtofs_glo_2ds_f000_ice.argo.nc
/lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/prep/rtofs/rtofs.20230530/argo/
rtofs_glo_2ds_f000_ice.argo.nc
/lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/prep/rtofs/rtofs.20230530/argo/
ice_mask.nc -type data -mask_field name="ice_coverage"; level="(0,*,*)";
-thresh lt0.15 -name ice_mask
DEBUG 1: Input Grid:
 /lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/prep/rtofs/rtofs.20230530/argo/
rtofs_glo_2ds_f000_ice.argo.nc
DEBUG 1: Mask File:
/lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/prep/rtofs/rtofs.20230530/argo/
rtofs_glo_2ds_f000_ice.argo.nc
DEBUG 2: Parsed Input Grid:     LatLon (3600 x 1800)
DEBUG 2: Parsed Mask Grid:      LatLon (3600 x 1800)
ERROR  :
ERROR  : write_netcdf() -> trouble opening output file
/lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/prep/rtofs/rtofs.20230530/argo/
ice_mask.nc

I think you were running as user: shelley.melchior(32081), but you tried to
write to emc.vpppg's noscrub directory, which you don't have permission to
write.  Please check your COMINfcst setup.  It should be the same as all
other runs (e.g., ghrsst, smap, smos etc.).  If other runs had no problems,
use the same setup.

Thanks,
Gwen

**Comment by ShelleyMelchior-NOAA** [comment]
I pointed all COMIN to the EVS para since I do not have enough run history. Here are my settings in the ecf file for COMINs and COMOUTs:
`#export COMIN=/lfs/h2/emc/vpppg/noscrub/$USER/$NET/${evs_ver}`
`export COMIN=/lfs/h2/emc/vpppg/noscrub/emc.vpppg/$NET/${evs_ver}`
`export COMINobs=/lfs/h1/ops/dev/dcom`
`#export COMINfcst=/lfs/h2/emc/vpppg/noscrub/$USER/$NET/${evs_ver}/prep/$COMPONENT`
`export COMINfcst=/lfs/h2/emc/vpppg/noscrub/emc.vpppg/$NET/${evs_ver}/prep/$COMPONENT`
`export COMINclimo=/lfs/h2/emc/vpppg/noscrub/emc.vpppg/verification/EVS_fix/climos/$COMPONENT`
`export COMOUT=/lfs/h2/emc/vpppg/noscrub/$USER/$NET/${evs_ver}`
`export COMOUTfinal=$COMOUT/$STEP/$COMPONENT/$COMPONENT.$VDATE`
Please advise if this is not correct.

**Comment by GwenChen-NOAA** [comment]
You changed COMINfcst from $USER to emc.vpppg, but you ran it
as shelley.melchior(32081) and can't write to emc.vpppg, causing the
errors.  It would be fine if you (1) run it as emc.vpppg, or (2) copy the
rtofs prep directory to your noscrub, such as:

> cp -rv
/lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/prep/rtofs/rtofs.2023052?
/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/prep/rtofs

Then set:
export COMINfcst=/lfs/h2/emc/vpppg/noscrub/*$USER*
/$NET/${evs_ver}/prep/$COMPONENT

And run it as shelley.melchior(32081).

Gwen

**Comment by ShelleyMelchior-NOAA** [comment]
Ok, so that is my confusion. I figured any COMIN is only a space to read from. I didn't realize COMIN is also a space to be writing into? Shouldn't the processing only be writing into COMOUT variables?

**Comment by ShelleyMelchior-NOAA** [comment]
I think this is something fundamental that needs to be changed in your processing. We should not be writing into any COMIN space. COMIN is reserved for reading from. COMOUT is reserved for writing into.

**Comment by ShelleyMelchior-NOAA** [comment]
@GwenChen-NOAA , I will do your recommendation for the sake of this PR and getting it moving along. But this might be a feature worth modifying by Samira for EVSv2. That or I just need a better understanding of why stats processing is writing into prep.

**Comment by GwenChen-NOAA** [comment]
@Shelley Melchior - NOAA Affiliate ***@***.***> , no stats
are written into prep.  All stats are written into COMOUTsmall and
COMOUTfinal directories according to the guidelines.  The files write into
the prep directory are dynamic mask files, which are generated daily based
on sea ice concentration.  The prep directory is an ideal place to put the
mask files.  This directory structure has been reviewed many times and
approved.  Please take some time to understand what the code is doing
before jumping into conclusion.

Thanks,
Gwen

**Comment by PerryShafran-NOAA** [comment]
Gwen - thank you for that important information.  We were unaware that the stats was writing data to the prep directory or if that was even OK, so what was happening seemed rather strange to me as well.  But now that we know, we will proceed by copying the prep directory into one of our own directories so this doesn't become a problem.  

Thanks!

Perry

**Comment by ShelleyMelchior-NOAA** [comment]
Thank you for the explanation, Gwen. I was definitely trying to understand the code which is why I was perplexed because the construct RTOFS has to use for dynamic mask files is something I hadn't yet come across in the various PRs I have tested, approved, and merged.

(NOTE: To read with the intended formatting, I suggest reading this update in the github interface.)

### STATS
I have cp'd the EVS para prep rtofs directories to my $USER vpppg noscrub. I have run all stats processing. The COMOUTs (small and final) populate. However, I see one problem that I suggest be corrected. The rsync to ARCHevs at the end of the exevs scripts did not work because ARCHevs was never created. Rather than rely on a user to create ARCHevs, the code should do that automatically. I suggest adding `mkdir -p $ARCHevs` right after the export in the ecf scripts **OR** right before its usage in the exevs scripts.

From `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_rtofs_smap_stats.o74030056` you will see the rsync error ~line 1986:
`rsync: [Receiver] mkdir "/lfs/h2/emc/vpppg/noscrub/shelley.melchior/stat_archive/RTOFS" failed:`
` No such file or directory (2)`
`rsync error: error in file IO (code 11) at main.c(758) [Receiver=3.2.2]`

I have modified the ecf scripts in my clone of your fork to include `mkdir -p $ARCHevs` just after the export of ARCHevs. I re-ran all of the stats processing. The ARCHevs directory is created and properly populated.
`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/stat_archive/RTOFS/rtofs.20230530/`
Incidentally, @PerryShafran-NOAA , I notice the EVS para also does not have the stat_archive/ directory. Is that intentional? When Gwen modifies the code in her develop and this PR is approved and merged into the auth EVS develop, you can get that change into the EVS para and the stat_archive/ dir will be created for emc.vpppg.

The ndbc stats job took ~3.5 hrs to complete. I notice the specified walltime is 6 hours.
The argo stats job took a bit of time as well (~1 hr). I notice the specified walltime is 3 hours. 
Given the lengthy walltime requests, I presume this is known and expected for both argo and ndbc.

### PLOTS
The plots jobs also each have walltime requests of 6 hours. But the headline plots job completed in ~5 mins. It may have run so quickly because there are only 2 days to plot? Some un-tarred files are available on RZDM for quick viewing:
https://ftp.emc.ncep.noaa.gov/users/smelchior/EVS/rtofs1/lfs/h2/emc/ptmp/shelley.melchior/evs/v1.0/plots/rtofs.20230530/headline/
To get more data, I re-ran pointing to the EVS para stats files. This is another strange feature, though. COMOUTfinal is what is used to define where to retrieve (read) the stats files. Reading should come from a COMIN variable. Usage of COMOUTfinal for stats processing is perfect, but for plots processing COMOUTfinal should be renamed to something like COMINstats.
Also worth noting, when pointing to the EVS para stats files, the headline job completed in 6.5 minutes. The walltime for headline can be reduced.
Some un-tarred files are available on RZDM for quick viewing:
https://ftp.emc.ncep.noaa.gov/users/smelchior/EVS/rtofs2/lfs/h2/emc/ptmp/shelley.melchior/evs/v1.0/plots/rtofs.20230530/headline/

The standard plots job took ~45 mins when using my local $USER stats. 
Some un-tarred files are available on RZDM for quick viewing:
https://ftp.emc.ncep.noaa.gov/users/smelchior/EVS/rtofs3/lfs/h2/emc/ptmp/shelley.melchior/evs/v1.0/plots/rtofs.20230530/
To get more data, I re-ran pointing to the EVS para stats files (modifying COMOUTfinal as mentioned above). The re-run took ~60 mins. The walltime can be reduced for the standard plots job, too.
Some un-tarred files are available on RZDM for quick viewing:
https://ftp.emc.ncep.noaa.gov/users/smelchior/EVS/rtofs4/lfs/h2/emc/ptmp/shelley.melchior/evs/v1.0/plots/rtofs.20230530/

### SUMMARY
Please see the log output and the stats and plots output products at the paths shown below:
logfiles: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out`
`COMOUT=/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0`
code: `HOMEevs=/lfs/h2/emc/vpppg/noscrub/shelley.melchior/githubwkspc/forks/GC/EVS`

To my knowledge, everything intended for the PR works correctly.

Before I can merge:
- sync `GwenChen-NOAA:develop` fork w/ auth EVS develop

Suggested changes to take care of now, otherwise they will require a subsequent PR:
- change EVS to HOMEevs
- include `mkdir -p $ARCHevs` either in ecf or in exevs scripts
- reduce walltime for plots ecf 

I am not wild about using COMOUTfinal as an input variable to read in stats files for plots processing. It works but is misleading and does not conform to convention. I suggest
- change COMOUTfinal to COMINstats for plots processing

**Comment by GwenChen-NOAA** [comment]
@Shelley Melchior - NOAA Affiliate ***@***.***> , thanks for
your thorough review!  I made the following changes to the code:

1. changed EVS to HOMEevs
2. changed COMOUTfinal to COMINstats for plots processing
3. reduced walltime for plots ecf
4. removed rsync to ARCHevs in exevs scripts. This function is not needed
for EVS because it will have another way to archive to HPSS. I have put
notes in previous PRs to ignore it. It's about time to comment it out.

Please merge the latest version of code at
https://github.com/GwenChen-NOAA/EVS/tree/develop.  It has synced with the
auth EVS repo.

Thanks,
Gwen

**Comment by GwenChen-NOAA** [comment]
Missing one item:

5. moved individual plots directory (rtofs.$VDATE) from $DATA/$STEP to
$DATA/$STEP/$COMPONENT to be consistent with COMOUTplots.

Gwen

**Comment by ShelleyMelchior-NOAA** [comment]
Sounds great, @GwenChen-NOAA . I will do a quick review of the code for anything else and then do the merge.

**Review by ShelleyMelchior-NOAA — APPROVED** [review]
Code changes look good and are expected for the realm of this PR. Approved.

**Pull request merged**
